Make `static_show()` print valid identifiers

NHDaly opened the pull request:
This PR makes a few separate changes to `static_show()`, to bring it more in line with julia's `show()`, and to fix cases where it was printing non-valid code.
 - Use `var""` syntax in static_show to print names containing invalid characters as valid identifiers.
 - Start printing function instances by instance name, instead of via a non-existent type constructor.


Commit messages follow:

---------------------------

Use `var""` syntax in static_show to print valid identifiers

Previously, `static_show` was printing invalid identifiers for
`TypeVar`s that contain a `#`, such as those produced by `gensym()`.

For example:
```julia
julia> println(static_shown(Vector{<:Bool}))
Array{#s5, 1} where #s5<:Bool

julia> eval(Meta.parse(static_shown(Vector{<:Bool})))
ERROR: syntax: incomplete: premature end of input
```
After this commit, this is printed as a valid identifier:
```julia
julia> println(static_shown(Vector{<:Bool}))
Array{var"#s5", 1} where var"#s5"<:Bool

julia> eval(Meta.parse(static_shown(Vector{<:Bool})))
Vector{var"#s5"} where var"#s5"<:Bool (alias for Array{var"#s5", 1} where var"#s5"<:Bool)
```
As you can see, this mirrors the printing done from julia's `show()`.

This commit also applies the same technique to simplify the printing for
Type names that contain a `#`, simply for asthetic reasons and
consistency with julia `show()`, by reusing the above functionality.

Before:
```julia
julia> struct var"#X#" x::Int end

julia> println(static_shown(var"#X#"))
getfield(Main, Symbol("#X#"))

julia> println(static_shown(var"#X#"(0)))
getfield(Main, Symbol("#X#"))(x=0)
```
After:
```julia
julia> struct var"#X#" x::Int end

julia> println(static_shown(var"#X#"))
Main.var"#X#"

julia> println(static_shown(var"#X#"(0)))
Main.var"#X#"(x=0)
```

-------------------------------

Make functions print as valid julia expressions.

Previously, function instances would be printed in `static_show()` as
elements of a `DataType`, with no special printing for `<:Function`s.
This led to printing them via _invalid_ syntax, since there is no empty
constructor for `T<:Function`.

Before:
```julia
julia> g() = 2
g (generic function with 1 method)

julia> println(static_shown(g))
typeof(Main.g)()

julia> eval(Meta.parse(static_shown(g)))
ERROR: MethodError: no method matching typeof(g)()
```

After:
```julia
julia> g() = 2
g (generic function with 1 method)

julia> println(static_shown(g))
Main.g

julia> eval(Meta.parse(static_shown(g)))
g (generic function with 1 method)
```

I chose to keep the Module prefix on the functions, so that they will
be valid julia expressions that you can enter at the REPL to reproduce
the function instance, even though this is a break from the julia
`show()` behavior.

One caveat is that this is still not correct for anonymous functions,
but I'm also not really sure what would be a better way to print
anonymous functions... I think this is _probably_ better than it was
before, but very open to suggestions for how to improve it.

Before:
```julia
julia> l = (x,y)->x+y

julia> println(static_shown(l))
getfield(Main, Symbol("#3#4"))()

julia> eval(Meta.parse(static_shown(l)))
ERROR: MethodError: no method matching var"#3#4"()
```
After:
```julia
julia> l = (x,y)->x+y

julia> println(static_shown(l))
Main.var"#3"

julia> eval(Meta.parse(static_shown(l)))
ERROR: UndefVarError: #3 not defined
```

-----------------

Extend `static_show()` escaping to any illegal identifier via var"".

Previously, it was only escaping names with a `#` in them, but now it
will escape all names that are illegal identifiers.

Before:
```julia
julia> struct var"a%b+c" x::Int end

julia> println(static_shown(var"a%b+c"))
Main.a%b+c

julia> println(static_shown(var"a%b+c"(1)))
Main.a%b+c(x=1)
```
After:
```julia
julia> struct var"a%b+c" x::Int end

julia> println(static_shown(var"a%b+c"))
Main.var"a%b+c"

julia> println(static_shown(var"a%b+c"(1)))
Main.var"a%b+c"(x=1)
```

NHDaly committed:
Use `var""` syntax in static_show to print valid identifiers

Previously, `static_show` was printing invalid identifiers for
`TypeVar`s that contain a `#`, such as those produced by `gensym()`.

For example:
```julia
julia> println(static_shown(Vector{<:Bool}))
Array{#s5, 1} where #s5<:Bool

julia> eval(Meta.parse(static_shown(Vector{<:Bool})))
ERROR: syntax: incomplete: premature end of input
```
After this commit, this is printed as a valid identifier:
```julia
julia> println(static_shown(Vector{<:Bool}))
Array{var"#s5", 1} where var"#s5"<:Bool

julia> eval(Meta.parse(static_shown(Vector{<:Bool})))
Vector{var"#s5"} where var"#s5"<:Bool (alias for Array{var"#s5", 1} where var"#s5"<:Bool)
```
As you can see, this mirrors the printing done from julia's `show()`.

This commit also applies the same technique to simplify the printing for
Type names that contain a `#`, simply for asthetic reasons and
consistency with julia `show()`, by reusing the above functionality.

Before:
```julia
julia> struct var"#X#" x::Int end

julia> println(static_shown(var"#X#"))
getfield(Main, Symbol("#X#"))

julia> println(static_shown(var"#X#"(0)))
getfield(Main, Symbol("#X#"))(x=0)
```
After:
```julia
julia> struct var"#X#" x::Int end

julia> println(static_shown(var"#X#"))
Main.var"#X#"

julia> println(static_shown(var"#X#"(0)))
Main.var"#X#"(x=0)
```

NHDaly committed:
Make functions print as valid julia expressions.

Previously, function instances would be printed in `static_show()` as
elements of a `DataType`, with no special printing for `<:Function`s.
This led to printing them via _invalid_ syntax, since there is no empty
constructor for `T<:Function`.

Before:
```julia
julia> g() = 2
g (generic function with 1 method)

julia> println(static_shown(g))
typeof(Main.g)()

julia> eval(Meta.parse(static_shown(g)))
ERROR: MethodError: no method matching typeof(g)()
```

After:
```julia
julia> g() = 2
g (generic function with 1 method)

julia> println(static_shown(g))
Main.g

julia> eval(Meta.parse(static_shown(g)))
g (generic function with 1 method)
```

I chose to keep the Module prefix on the functions, so that they will
be valid julia expressions that you can enter at the REPL to reproduce
the function instance, even though this is a break from the julia
`show()` behavior.

One caveat is that this is still not correct for anonymous functions,
but I'm also not really sure what would be a better way to print
anonymous functions... I think this is _probably_ better than it was
before, but very open to suggestions for how to improve it.
Before:
```julia
julia> l = (x,y)->x+y

julia> println(static_shown(l))
getfield(Main, Symbol("#3#4"))()

julia> eval(Meta.parse(static_shown(l)))
ERROR: MethodError: no method matching var"#3#4"()
```
After:
```julia
julia> l = (x,y)->x+y

julia> println(static_shown(l))
Main.var"#3"

julia> eval(Meta.parse(static_shown(l)))
ERROR: UndefVarError: #3 not defined
```

NHDaly commented:
This addresses `Case 2` in this issue about emitting invalid precompile statements: https://github.com/JuliaLang/julia/pull/28808

NHDaly committed:
Extend `static_show()` escaping to any illegal identifier via var"".

Previously, it was only escaping names with a `#` in them, but now it
will escape all names that are illegal identifiers.

Before:
```julia
julia> struct var"a%b+c" x::Int end

julia> println(static_shown(var"a%b+c"))
Main.a%b+c

julia> println(static_shown(var"a%b+c"(1)))
Main.a%b+c(x=1)
```
After:
```julia
julia> struct var"a%b+c" x::Int end

julia> println(static_shown(var"a%b+c"))
Main.var"a%b+c"

julia> println(static_shown(var"a%b+c"(1)))
Main.var"a%b+c"(x=1)
```

Sacha0 reviewed (commented):
First commit looks great! :)

Sacha0 reviewed (commented):
Second commit looks great! --- modulo a couple things I don't entirely grok; let's chat offline :).

Sacha0 commented on src/julia.h:
```diff
@@ -1105,6 +1105,7 @@ static inline int jl_is_layout_opaque(const jl_datatype_layout_t *l) JL_NOTSAFEP
 #define jl_is_svec(v)        jl_typeis(v,jl_simplevector_type)
 #define jl_is_simplevector(v) jl_is_svec(v)
 #define jl_is_datatype(v)    jl_typeis(v,jl_datatype_type)
+#define jl_is_function(v)    jl_typeis(v,jl_function_type)
```
Ah, modulo this line --- did you mean to consume it later in the commit, perhaps in the `jl_function_type` branch?

NHDaly replied:
Oops, you're totally right. Ultimately i did not need this. Thanks! :)

Sacha0 reviewed (commented):
Third commit looks great! :)

Sacha0 reviewed (approved):
Confusion resolved. Looks great! :)

Sacha0 commented:
Absent objections or requests for time, I plan to merge this pull request on Saturday morning PT. Best! :)

NHDaly committed:
Add comments to some of the static_show blocks.

I was particularly confused when reading the code between the block for
instances of DataTypes and instances of types that are DataTypes, so I
hope these comments will help future readers be less confused. :)

JeffBezanson commented on src/rtutils.c:
```diff
@@ -604,6 +604,42 @@ JL_DLLEXPORT jl_value_t *jl_argument_datatype(jl_value_t *argt JL_PROPAGATES_ROO
     return (jl_value_t*)dt;
 }
 
+int is_globfunction(jl_value_t *v, jl_datatype_t *dv, jl_sym_t **globname_out) {
```
Please add a newline before the `{` to match style. This could also be declared `static`.

NHDaly replied:
Oops! :) Thanks. done in latest push.

JeffBezanson commented on src/rtutils.c:
```diff
@@ -604,6 +604,42 @@ JL_DLLEXPORT jl_value_t *jl_argument_datatype(jl_value_t *argt JL_PROPAGATES_ROO
+
+    char *sn = jl_symbol_name(name);
+    int hidden = 0;
+    if (!jl_is_identifier(sn) || jl_is_operator(sn)) {
```
Is the `isoperator` check really needed? For example this gives `typeof(Base.:(var"+"))` instead of `typeof(Base.:(+))`.

NHDaly replied:
Oops! thanks Jeff. I really think i meant to make this `if (!(jl_is_identifier(sn) || jl_is_operator(sn)))`. Because already `jl_is_identifier("+")` returns false, so i think i added this specifically to _allow_ operators to be okay, but i screwed up the parens. That's what i get for coding late at night 😅 thanks - this is a great catch!

```
julia> @ccall jl_is_identifier("+"::Cstring)::Bool
false

julia> @ccall(jl_is_identifier("+"::Cstring)::Bool) || @ccall(jl_is_operator("+"::Cstring)::Bool)
true
```

i've amended the commit to fix this, and also added an expected output test for this case. 👍  thanks!

NHDaly replied:
This is now fixed. Does this sound right to you? Thanks.
